### PR TITLE
python scan performance improvements

### PR DESCRIPTION
**Pull request by yaronschwimmer**
- [X] Ready for review
- [X] Follows CONTRIBUTING rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?
Fix some performance issues with python app scan:
1. Handle cyclic dependencies
2. Add pre-processing of metadata files to avoid sorting the same files over and over

### CODE REVIEW

File: test/system/application-scans/python/pip.spec.ts [hunk 1]
```diff
@@ -15,7 +15,7 @@ describe("pip application scan", () => {
   });
 
   it("should handle --exclude-app-vulns with string and boolean value", async () => {
-    const fixturePath = getFixture("docker-archives/docker-save/php.tar");
+    const fixturePath = getFixture("docker-archives/docker-save/pip.tar");
```
Comment (RotemS) [inline test/system/application-scans/python/pip.spec.ts:18]: On purpose?

File: lib/analyzer/applications/python/pip.ts [hunk 2]
```diff
@@ -12,13 +12,14 @@ import { AppDepsScanResultWithoutTarget, FilePathToContent } from "../types";
 const debug = Debug("snyk");
 class PythonDepGraphBuilder {
   private requirements: PythonRequirement[];
-  private metadata: PythonPackage[];
+  private metadata: { [name: string]: PythonPackage[] };
```
Comment (RotemS) [inline lib/analyzer/applications/python/pip.ts:15]: Worth a specific type?

File: test/lib/analyzer/python-pip-analyzer.spec.ts [hunk 3]
```diff
@@ -115,4 +115,14 @@ describe("pip analyzer", () => {
...
   });
+
+  it("handles cyclic dependencies", async () => {
+    const filePathToContent = prepareFiles("cyclic");
+    const res = await pipFilesToScannedProjects(filePathToContent);
+    expect(res[0].identity).toMatchObject({
+      type: "pip",
+      targetFile: "/app/requirements.txt",
+    });
+    expect(res[0].facts[0].data._depPkgsList).toHaveLength(8);
```
Comment (RotemS) [inline test/lib/analyzer/python-pip-analyzer.spec.ts:126]: Is it worth to add a check here that the graph's structure is correct as well?

File: test/system/application-scans/python/pip.spec.ts (hunk 1 above)
Comment (yaronschwimmer) [reply test/system/application-scans/python/pip.spec.ts:18]: yes. it was a copy-paste error

File: lib/analyzer/applications/python/pip.ts (hunk 2 above)
Comment (yaronschwimmer) [reply lib/analyzer/applications/python/pip.ts:15]: sure. why not

**Review by RotemS — APPROVED** [review]

**Pull request merged**